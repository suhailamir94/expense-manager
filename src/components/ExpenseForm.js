import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const now = moment();

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.onDescriptionChange = this.onDescriptionChange.bind(this);
    this.onAmountChange = this.onAmountChange.bind(this);
    this.onNoteChange = this.onNoteChange.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? props.expense.amount.toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: "",
    };
  }
  onDescriptionChange(e) {
    const description = e.target.value;
    this.setState(prevState => ({ ...prevState, description }));
  }
  onNoteChange(e) {
    const note = e.target.value;
    this.setState(prevState => ({ ...prevState, note }));
  }

  onAmountChange(e) {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}$/))
      this.setState(prevState => ({ ...prevState, amount }));
  }

  onDateChange(createdAt) {
    if (createdAt)
      this.setState(() => ({
        createdAt,
      }));
  }

  onFocusChange({ focused }) {
    this.setState(() => ({ calenderFocused: focused }));
  }
  onSubmit(e) {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Please provide description and amount" }));
    } else {
      this.state.error && this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: this.state.amount,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Descreption"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt} // momentPropTypes.momentObj or null
            onDateChange={this.onDateChange} // PropTypes.func.isRequired
            focused={this.state.calenderFocused} // PropTypes.bool
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
