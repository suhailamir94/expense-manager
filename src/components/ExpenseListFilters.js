import React from "react";
import { connect } from "react-redux";

import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from "../actions/filters";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);
    this.onDatesChange = this.onDatesChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onTextChange = this.onTextChange.bind(this);
    this.onSortChange = this.onSortChange.bind(this);
    this.state = { calenderFocused: null };
  }

  onDatesChange({ startDate, endDate }) {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange(calenderFocused) {
    this.setState(() => ({ calenderFocused }));
  }

  onTextChange(e) {
    this.props.setTextFilter(e.target.value);
  }
  onSortChange(e) {
    e.target.value === "amount"
      ? this.props.sortByAmount()
      : this.props.sortByDate();
  }
  render() {
    return (
      <div className="content-container">
        <div class="input-group">
          <div class="input-group__item">
            <input
              className="text-input"
              type="text"
              placeholder="Search Expenses"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div class="input-group__item">
            <select
              className="select"
              defaultValue={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div class="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calenderFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
