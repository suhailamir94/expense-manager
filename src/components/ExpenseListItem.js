import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <h3>
      <Link to={`/edit/${id}`}>{description}</Link>
    </h3>
    <p>
      &#x20B9;{amount} - {moment(createdAt).format("MMMM Do, YYYY")}
    </p>
  </div>
);

export default ExpenseListItem;
