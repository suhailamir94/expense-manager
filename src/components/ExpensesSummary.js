import React from "react";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import totalExpense from "../selectors/expenses-total";

export const ExpensesSummary = ({ expenseCount, expenseTotal }) => {
  const expenseWord = expenseCount ? "expenses" : "expense";
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling &#x20B9;{expenseTotal}
      </h1>
    </div>
  );
};
const mapStateToProps = state => {
  const expenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: expenses.length,
    expenseTotal: totalExpense(expenses),
  };
};
export default connect(mapStateToProps)(ExpensesSummary);
