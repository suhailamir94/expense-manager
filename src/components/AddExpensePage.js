import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

export const AddExpensePage = props => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      onSubmit={expense => {
        props.startAddExpense(expense); // this onSubmit is passed to props by the mapDispatchToProps method defined below
        // this is how we redirect from one page to another through JS in React Router
        props.history.push("/dashboard");
      }}
    />
  </div>
);

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(addExpense(expense)),
});
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
