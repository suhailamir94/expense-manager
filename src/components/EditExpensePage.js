import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export const EditExpensePage = props => {
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={expense => {
          props.onSubmit(props.expense.id, expense);
          // this is how we redirect from one page to another through JS in React Router
          props.history.push("/");
        }}
      />
      <button
        onClick={() => {
          props.onClickRemove(props.expense.id);
          // props.dispatch(removeExpense({ id: props.expense.id }));
          props.history.push("/");
        }}
      >
        REMOVE
      </button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  onSubmit: (id, expense) => dispatch(editExpense(id, expense)),
  onClickRemove: id => dispatch(removeExpense({ id })),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
