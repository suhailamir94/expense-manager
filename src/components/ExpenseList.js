import React from "react";
import { connect } from "react-redux";

import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = props => {
  return (
    <div>
      <h2>Expense List</h2>
      {props.expenses.length ? (
        props.expenses.map(expense => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      ) : (
        <p>No Expenses Found!!</p>
      )}
    </div>
  );
};
const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};
export default connect(mapStateToProps)(ExpenseList);
