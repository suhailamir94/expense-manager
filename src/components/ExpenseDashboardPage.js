import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
  <div className="container-column">
    <div className="container">
      <ExpenseListFilters />
      <ExpensesSummary />
      <ExpenseList />
    </div>
  </div>
);
export default ExpenseDashboardPage;
