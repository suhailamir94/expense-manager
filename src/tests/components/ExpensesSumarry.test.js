import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";

test("should render ExpenseSummary with 1 expense correctly", () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={1} expenseTotal={235} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary with multiple expenses correctly", () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={10} expenseTotal={2350} />
  );
  expect(wrapper).toMatchSnapshot();
});
