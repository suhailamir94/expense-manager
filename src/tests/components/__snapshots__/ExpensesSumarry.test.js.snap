// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with 1 expense correctly 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expenses
     totalling ₹
    235
  </h1>
</div>
`;

exports[`should render ExpenseSummary with multiple expenses correctly 1`] = `
<div>
  <h1>
    Viewing 
    10
     
    expenses
     totalling ₹
    2350
  </h1>
</div>
`;
